Python/pytime.c

```diff
@@ -700,29 +721,58 @@ pymonotonic(_PyTime_t *tp, _Py_clock_info_t *info, int raise)
 
 #elif defined(__APPLE__)
     static mach_timebase_info_data_t timebase;
-    uint64_t time;
+    static uint64_t t0 = 0;
+    uint64_t ticks;
 
     if (timebase.denom == 0) {
         /* According to the Technical Q&A QA1398, mach_timebase_info() cannot
            fail: https://developer.apple.com/library/mac/#qa/qa1398/ */
         (void)mach_timebase_info(&timebase);
-    }
 
-    time = mach_absolute_time();
+        /* Sanity check: should never occur in practice */
+        if (timebase.numer < 1 || timebase.denom < 1) {
+            PyErr_SetString(PyExc_RuntimeError,
+                            "invalid mach_timebase_info");
+            return -1;
+        }
 
-    /* apply timebase factor */
-    time *= timebase.numer;
-    time /= timebase.denom;
+        /* Sanity check: make sure that ticks * timebase.numer cannot overflow
+           in _PyTime_MulDiv(), with ticks < timebase.denom.
 
-    *tp = time;
+           Known time bases:
+
+           * always (1, 1) on Intel
+           * (1000000000, 33333335) or (1000000000, 25000000) on PowerPC
+
+           None of these time bases can overflow with 64-bit _PyTime_t.
+
+           Check also that timebase.numer and timebase.denom can be casted to
+           _PyTime_t. */
+        if (timebase.numer > _PyTime_MAX / timebase.denom
+            || timebase.denom > _PyTime_MAX) {
```

Comment:
`timebase.denom > _PyTime_MAX` is redundant. If `timebase.denom > _PyTime_MAX` then `_PyTime_MAX / timebase.denom == 0`.

Reply:
While it's correct, I prefer to keep the test to be very explicit. The test is only run once, it doesn't matter.

Reply:
Maybe than add the explicit test `timebase.numer > _PyTime_MAX`?

Additional `timebase.denom > _PyTime_MAX` doesn't add clarity to me. It rather make me puzzling and spending a minute for figuring out why this test is added explicitly.

Reply:
> Maybe than add the explicit test timebase.numer > _PyTime_MAX?

Done